### ISSUE TITLE
Optimization: Modularize asset kit for buildings

### ISSUE THREAD

**Issue by madjin** [NEON-BUIDL/DAOtown#12]
![image](https://github.com/NEON-BUIDL/DAOtown/assets/32600939/3e474873-5695-41b3-9925-dfebf3bb6575)

To make a bigger world, we need to modularize assets so that we can use gltf-extension for instancing that hyperfy now supports. Here is an example architecture kit from mozilla hubs: https://hubs.mozilla.com/docs/spoke-architecture-kit.html

![image](https://github.com/NEON-BUIDL/DAOtown/assets/32600939/b180dd83-d044-49e9-836e-da322682b65a)

This way we can save on draw calls. Things we want to reuse:

- cactus
- payphone
- walls / building parts
- signage (maybe its 1 object joined)
- roads / railroads
- carriage wagons
- props
  - barrels
  - rope
  - fencing
  - tables / chairs

**Comment by madjin** [comment]
## Proposal: Builder App

There's a ton of building asset kits already: https://sketchfab.com/search?q=building+kit&type=models

Instead of making a new asset kit, reinventing the wheel, I think our efforts would be better spent tapping into the power of real-time collaborative world building that Hyperfy enables with an app that allows us to load asset kits and use them with poweruser keyboard / mouse shortcuts.

- asset kit loadable from json
  - name
  - thumbnail
  - path
  - stats
  
User should be able to drag and drop into the world, and have snapping so that pieces can fit perfectly next to each other. This can be done with some keyboard shortcuts as well:

- Bring up the asset kit with 1 key (right mouse click or tab)
- 1-4 on keyboard to change factor
  - snap movement by 1, 0.1, 0.01, 0.001 
- Q and E to scroll / rotate
- Realtime raycast objects to surface

We'll be able to tap into the power of open standards like this gltf-extension for GPU instancing, making bigger worlds that are more performant, and other platforms that utilize this extension will be able to load these worlds in and be performant as well.

**Comment by madjin** [comment]
![image](https://github.com/NEON-BUIDL/DAOtown/assets/32600939/668058f0-5ef4-409a-971c-e6359b0ae731)
![image](https://github.com/NEON-BUIDL/DAOtown/assets/32600939/3602cda7-2e77-492e-93fa-2bee9c7c6548)

**Comment by madjin** [comment]
tldr steps to get instancing to work
- use duplicate in blender to share geom/mats
- use gltf-report/gltf-transform instance() function
- preview in gltf-viewer (gltf-report doesnt have previews for instancing)

can use an asset pallet (see https://github.com/M3-org/base-meshes) to test with, and go from there

**Comment by madjin** [comment]
there's a cost involved in making something instanced because in threejs you have to specify the amount of instances
so if you have 100 completely different models all with "instancing" enabled but theres only 1 of each, that would be horrible

need to be smart about how to go about it in a way that makes sense, like some kind of heuristic that counts how many times an object spawned and shove things >2 into instance

perhaps could add ?instance=true in the asset pallete url so the platform knows to instance? or detect objects after spawning in some kind of manual optimization step for worlds
